sample/transforms.go

```diff
@@ -26,44 +26,53 @@ func (h *tokenHeap) Pop() any {
 }
 
 // temperature applies scaling to the logits
-func temperature(ts []token, temp float32) []token {
+func temperature(ts *[]token, temp float32) {
+	if ts == nil {
+		return
+	}
+
 	// Ensure temperature clipping near 0 to avoid numerical instability
 	temp = max(temp, 1e-7)
-	for i := range ts {
-		ts[i].value = ts[i].value / temp
+	for i := range *ts {
+		(*ts)[i].value = (*ts)[i].value / temp
 	}
-	return ts
 }
 
 // softmax applies normalization to the logits
-func softmax(ts []token) []token {
+func softmax(ts *[]token) {
+	if ts == nil {
+		return
+	}
+
 	// Find max logit for numerical stability
 	maxLogit := float32(math.Inf(-1))
-	for _, t := range ts {
+	for _, t := range *ts {
 		if t.value > maxLogit {
 			maxLogit = t.value
 		}
 	}
 
 	// Compute exp(x - max)
 	var sum float32
-	for i, v := range ts {
-		ts[i].value = float32(math.Exp(float64(v.value - maxLogit)))
-		sum += ts[i].value
+	for i, v := range *ts {
+		(*ts)[i].value = float32(math.Exp(float64(v.value - maxLogit)))
+		sum += (*ts)[i].value
 	}
 
 	// exp(x - max) / sum(exp(x - max))
-	for i := range ts {
-		ts[i].value /= sum
+	for i := range *ts {
+		(*ts)[i].value /= sum
 	}
-
-	return ts
 }
 
 // topK limits the number of tokens considered to the k highest logits
-func topK(ts []token, k int) []token {
-	if k >= len(ts) || k <= 0 {
-		slices.SortFunc(ts, func(a, b token) int {
+func topK(ts *[]token, k int) {
```

Comment:
Add comment as to why it's passed in by pointer here. Something like the length of the slice is modified to k